jme3-core/src/main/java/com/jme3/anim/tween/action/Action.java

```diff
@@ -40,44 +105,61 @@ public void setLength(double length) {
         this.length = length;
     }
 
+    /**
+     * Retrieves the speedup factor applied by the layer for this action.
+     *
+     * @see Action#setSpeed(double) for detailed documentation
+     * @return the speed of frames.
+     */
     public double getSpeed() {
         return speed;
     }
 
+    /**
+     * Alters the speedup factor applied by the layer running this action.
+     * <p>
+     * Notes:
+     * <li> This factor controls the animation direction, if the speed is a positive value then the animation will run forward and vice versa. </li>
+     * <li> The speed factor gets applied, inside the {@link com.jme3.anim.AnimLayer}, on each interpolation step by this formula : time += tpf * action.getSpeed() * composer.globalSpeed. </li>
+     * <li> Default speed is 1.0, it plays the animation clips at their normal speed. </li>
+     * <li> Setting the speed factor to Zero will stop the animation, while setting it to a negative number will play the animation in a backward fashion. </li>
+     * </p>
+     * 
+     * @param speed the speed of frames.
+     */
     public void setSpeed(double speed) {
         this.speed = speed;
-        if( speed < 0){
+        if (speed < 0) {
             setForward(false);
         } else {
             setForward(true);
         }
     }
 
+    /**
+     * Retrieves the animation mask for this action.
+     * The animation mask controls which part of the model would be animated. A model part can be
+     * registered using a {@link com.jme3.anim.Joint}.
+     *
+     * @return the animation mask instance.
+     */
     public AnimationMask getMask() {
         return mask;
     }
 
+    /**
+     * Adjusts the animation mask for this action.
+     * The animation mask controls which part of the model would be animated. A model part can be
+     * registered using a {@link com.jme3.anim.Joint}.
+     *
+     * @param mask the animation mask instance.
+     */
     public void setMask(AnimationMask mask) {
```

Comment:
>This is the effect of setting it to null when also the AnimLayer has a null instance, it skips the interpolation (stops the animation):

Nope, please read the `if` statement carefully! It will skip it only if you specify a mask and that mask does not contain the bone.

Reply:
Oops, sorry.

Reply:
I still wonder what will happen if someone sets the mask to null, do you have any idea?

Reply:
When the mask is null it means that you want the animation to affect all the bones, if you want an animation to affect only a subset of bones, for example only the right hand then you specify a mask that only contains the joints from the right hand.

Reply:
Someone might want to create a sequence of multiple actions and wants each action to have a different mask (e.g. first one on the right hand and the second one on the head) then he must specify the mask manually.

Reply:
Will it get overwritten by the AnimLayer#update(float) to the mask specified by the AnimLayer?

Reply:
I mean setting the mask from here will have no effect if the animation layer will overwrite its value on each update.

Reply:
They can turn off "mask propagation" so it won't get overridden by AnimLayer. 

And now I feel stupid, as instead I could just check in the AnimLayer and apply the mask only if it is not null.

https://github.com/jMonkeyEngine/jmonkeyengine/blob/efbcd03e5806b9d33296e65e90ded19413dee804/jme3-core/src/main/java/com/jme3/anim/AnimLayer.java#L269-L271

Reply:
I don't know if the mask propagation flag will have an effect on this, it is better to discuss this part on the forums.

Reply:
My understanding of the architecture of this animation is that only the AnimLayer should take control of setting up the animation mask for the actions and all actions under a single Animation layer will run on the same bone, and this might be a limitation of this system.

Reply:
> or BlendAction anim layer will always override the mask unless they are wrapped inside a BaseAction and disable mask pro

What is special about `BaseAction` that stops the animation layer from overwriting the value of the animation mask?

I think this is literally a bad design, to specialize an entity to adjust the AnimationMask which is `AnimLayer`, then bleed the mask instance to be adjustable by raw Actions.

It's better to find a good alternative for the example "Running a sequence of actions with each action acting on a unique bone", for example, using CallableTweens to adjust the layer mask or something.

Reply:
Yeah, it is discussable. I also think that was the original idea of the original author but it has the limitation I mentioned above also it brokes the design because it needs to hijack the input tween and extract any BlendableAction from it so it can pass the mask, which is really wrong I think.

Also in the original code this was broken as well, and if you were creating a sequence action AnimLayer was never able to set the mask on the internal actions because  BaseAction was never forwarding the mask to them, which I fixed recently in the JME 3.6 release. (that was when I added mask propagation)

Reply:
>What is special about BaseAction that stops the animation layer from overwriting the value of the animation mask?

See this forum issue:

https://hub.jmonkeyengine.org/t/solved-actions-generated-with-actionsequence-affecting-other-layers/46797

Reply:
It feels a little bit messy. I don't know what I should do, if I documented this method as "Internal use only" it will be misleading; because it is usable for sub-actions when utilizing BaseAction in code, and if I documented it as being usable in all cases, it will be totally wrong; because that polymorphism at the BaseAction level changes the method activity entirely, I feel much like it's not polymorphism anymore!

I guess it's best to override this method and override its documentation for each implementation, what do you think about this?

Reply:
I think this method should never be called from BaseAction, AnimLayer will take care after the addition of mask propagation.

As for ClipAction, as you explained, this method can be called from a ClipAction if and only if this ClipAction is being wrapped inside a BaseAction and the "mask propagation" is turned off such that AnimLayer will never overwrite the newly written value.

So, overrding the `Action#setMask(...)` for ClipAction and changing the documentation for this specific implementation might be the best option.

And, keeping `Action#setMask(...)` marked as internal use only.

Reply:
I do not know, but I like the previous documentation. For example, in my case I do not use AnimLayer and AnimComposer, instead, I play actions using Lemur TweenAnimation and AnimationState and the set mask myself.

Reply:
Lemur utilizes `com.simsilica.lemur.anim.Tween` which is not the same as `com.jme3.anim.Tween`, I haven't seen any example utilizing the new animation system using Lemur utilities, may be you can elaborate more, I am feeling a little bit lost.

Reply:
I cloned TweenAnimation and replaced `com.simsilica.lemur.anim.Tween` with `com.jme3.anim.Tween` and that's it :)

Reply:
I don't know, but I feel it's much like a trivial code at least from my opinion, I guess the documentation job is to cover the engine components only.

Reply:
Here is an example 

https://hub.jmonkeyengine.org/t/questions-about-animcomposer/46256/2?u=ali_rs

Reply:
It seems like an issue that needs a fix, maybe you find a hack using Lemur Tweens, but that is an obvious system defect that is being ignored.

Reply:
Yeah, I see what you mean, I am fine with this as far as core devs are ok with this.

Reply:
I see, I am fine if other core devs are ok with that.

Reply:
Further discussions on forums: 
https://hub.jmonkeyengine.org/t/abstract-action-setmask-is-an-internal-use-only/46833